st2common/st2common/util/output_schema.py

```diff
@@ -53,22 +54,39 @@ def _validate_action(action_schema, result, output_key):
 
 
 def mask_secret_output(ac_ex, output_value):
-    if not output_value:
-        return output_value
-
-    if not isinstance(output_value, dict) and not isinstance(output_value, list):
+    # We only support output_schema validation when the output_value is a JSON object.
+    # Invididual keys of that object can be marked secret, but the entire output
+    # object cannot be marked as secret.
+    # FIXME: Should we support non-objects under output_key?
```

Comment:
We should. Is this something we can include in this PR to fix?

Reply:
I suppose, but that will increase the scope of this PR. I'll have to play with the schema to see how to allow more than just object.

Reply:
As in my other comment, yes, this should be in-scope for this PR. I misread the schema. I will push an update.

Reply:
I added support for recursively masking output whatever form the spec is in.
But we still only support top-level output_value being objects, so I left my `isinstance(..., Mapping)` check in place.

What would I do if output_value is an array? Apply the output_schema to each of the (assumed) objects in that array?
That's not how _validate_runner() and _validate_action() use the output_schema. So, other places would need to learn to support non-object output.